org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/JDTLanguageServer.java

```diff
@@ -203,7 +203,7 @@ public void didChangeConfiguration(DidChangeConfigurationParams params) {
 			}
 		}
 		if (preferenceManager.getClientPreferences().isCodeLensDynamicRegistrationSupported()) {
-			if (preferenceManager.getPreferences().isReferencesCodeLensEnabled()) {
+			if (preferenceManager.getPreferences().isCodeLensEnabled()) {
```

Comment:
Ok, so this is a bit trickier than expected ;-)

until now, whenever the referencesCodeLens preference was toggled, the capability was (un)registered accordingly, meaning the client would automatically adjust calling `document/codeLens`.

Now, if referencesCodeLens and implementationsCodeLens are both enabled, but you disable one of them, there's no unregistration event, meaning the client will keep displaying both links on the currently opened document.

We basically need to call (un)registerCapability, whenever one of the 2 preferences has changed, to force the client to adjust to the change